winpr/libwinpr/sspi/Negotiate/negotiate.c

```diff
@@ -72,45 +99,184 @@ const SecPkgInfoW NEGOTIATE_SecPkgInfoW = {
+	buf = negotiate_write_tlv(buf, tag, len, value);
+
+	return buf;
+}
+
+static BYTE* negotiate_read_tlv(BYTE* buf, BYTE* tag, ULONG* len, ULONG* bytes_remain)
+{
+	int len_bytes = 0;
+	*len = 0;
+
+	if (*bytes_remain < 2)
+		return NULL;
+	*bytes_remain -= 2;
+
+	*tag = *buf++;
+
+	if (*buf <= 0x7F)
+		*len = *buf++;
 	else
 	{
-		context->sspiA = &NTLM_SecurityFunctionTableA;
-		context->sspiW = &NTLM_SecurityFunctionTableW;
-		context->kerberos = FALSE;
+		len_bytes = *buf++ & 0x7F;
+		if (*bytes_remain < len_bytes)
+			return NULL;
+		*bytes_remain -= len_bytes;
+		for (int i = len_bytes - 1; i >= 0; i--)
+			*len |= *buf++ << (i * 8);
 	}
+
+	return buf;
 }
 
-static NEGOTIATE_CONTEXT* negotiate_ContextNew(void)
+static NEGOTIATE_CONTEXT* negotiate_ContextNew(NEGOTIATE_CONTEXT* init_context)
 {
-	NEGOTIATE_CONTEXT* context;
-	context = (NEGOTIATE_CONTEXT*)calloc(1, sizeof(NEGOTIATE_CONTEXT));
+	NEGOTIATE_CONTEXT* context = NULL;
 
+	context = calloc(1, sizeof(NEGOTIATE_CONTEXT));
 	if (!context)
 		return NULL;
 
-	context->NegotiateFlags = 0;
-	context->state = NEGOTIATE_STATE_INITIAL;
-	SecInvalidateHandle(&(context->SubContext));
-	negotiate_SetSubPackage(context, KERBEROS_SSP_NAME);
+	if (init_context->spnego)
+	{
+		init_context->mechTypes.pvBuffer = malloc(init_context->mechTypes.cbBuffer);
+		if (!init_context->mechTypes.pvBuffer)
+		{
+			free(context);
+			return NULL;
+		}
+	}
+
+	CopyMemory(context, init_context, sizeof(NEGOTIATE_CONTEXT));
+
 	return context;
 }
 
 static void negotiate_ContextFree(NEGOTIATE_CONTEXT* context)
 {
+	if (context->mechTypes.pvBuffer)
+		free(context->mechTypes.pvBuffer);
 	free(context);
 }
 
+static BOOL negotiate_oid_compare(const sspi_gss_OID_desc* oid1, const sspi_gss_OID_desc* oid2)
+{
+	return (oid1->length == oid2->length) && (memcmp(oid1->elements, oid2->elements, oid1->length) == 0);
+}
+
+static const char* negotiate_mech_name(const sspi_gss_OID_desc* oid)
+{
+	if (negotiate_oid_compare(oid, &spnego_OID))
+		return "SPNEGO (1.3.6.1.5.5.2)";
+	else if (negotiate_oid_compare(oid, &kerberos_OID))
+		return "Kerberos (1.2.840.113554.1.2.2)";
+	else if (negotiate_oid_compare(oid, &kerberos_wrong_OID))
+		return "Kerberos [wrong OID] (1.2.840.48018.1.2.2)";
+	else if (negotiate_oid_compare(oid, &ntlm_OID))
+		return "NTLM (1.3.6.1.4.1.311.2.2.10)";
+	else
+		return "Unknown mechanism";
+}
+
+static const Mech* negotiate_GetMechByOID(sspi_gss_OID_desc oid)
+{
+	if (negotiate_oid_compare(&oid, &kerberos_wrong_OID))
+	{
+		oid.length = kerberos_OID.length;
+		oid.elements = kerberos_OID.elements;
+	}
+
+	for (int i = 0; i < MECH_COUNT; i++)
+	{
+		if (negotiate_oid_compare(&oid, MechTable[i].oid))
+			return &MechTable[i];
+	}
+	return NULL;
+}
+
+static PSecHandle negotiate_FindCredential(MechCred* creds, const Mech* mech)
+{
+	for (int i = 0; i < MECH_COUNT; i++)
+	{
+		if (creds[i].mech == mech)
+		{
+			if (creds[i].valid)
+				return &creds[i].cred;
+			break;
```

Comment:
perhaps `return NULL;` would be clearer ?

Reply:
Ah, I see